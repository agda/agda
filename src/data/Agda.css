/* Aspects. */
.Agda .Comment       { color: #B22222 }
.Agda .Background    {}
.Agda .Markup        { color: #000000 }
.Agda .Keyword       { color: #CD6600 }
.Agda .String        { color: #B22222 }
.Agda .Number        { color: #A020F0 }
.Agda .Symbol        { color: #404040 }
.Agda .PrimitiveType { color: #0000CD }
.Agda .Pragma        { color: black   }
.Agda .Operator      {}
.Agda .Hole          { background: #B4EEB4 }

/* NameKinds. */
.Agda .Bound                  { color: black   }
.Agda .Generalizable          { color: black   }
.Agda .InductiveConstructor   { color: #008B00 }
.Agda .CoinductiveConstructor { color: #8B7500 }
.Agda .Datatype               { color: #0000CD }
.Agda .Field                  { color: #EE1289 }
.Agda .Function               { color: #0000CD }
.Agda .Module                 { color: #A020F0 }
.Agda .Postulate              { color: #0000CD }
.Agda .Primitive              { color: #0000CD }
.Agda .Record                 { color: #0000CD }

/* OtherAspects. */
.Agda .DottedPattern        {}
.Agda .UnsolvedMeta         { color: black; background: yellow         }
.Agda .UnsolvedConstraint   { color: black; background: yellow         }
.Agda .TerminationProblem   { color: black; background: #FFA07A        }
.Agda .IncompletePattern    { color: black; background: #F5DEB3        }
.Agda .Error                { color: red;   text-decoration: underline }
.Agda .TypeChecks           { color: black; background: #ADD8E6        }
.Agda .Deadcode             { color: black; background: #808080        }
.Agda .ShadowingInTelescope { color: black; background: #808080        }

/* Standard attributes. */
.Agda a { text-decoration: none }
.Agda a[href]:hover { background-color: #B4EEB4 }
.Agda [href].hover-highlight { background-color: #B4EEB4; }
