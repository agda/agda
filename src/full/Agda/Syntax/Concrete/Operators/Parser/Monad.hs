------------------------------------------------------------------------
-- | The parser monad used by the operator parser
------------------------------------------------------------------------

{-# LANGUAGE CPP           #-}
{-# LANGUAGE DeriveGeneric #-}

module Agda.Syntax.Concrete.Operators.Parser.Monad
  ( MemoKey(..), PrecedenceKey
  , Parser
  , parse
  , sat'
  , sat
  , doc
  , memoise
  , memoiseIfPrinting
  , grammar
  ) where

import Data.Hashable
import GHC.Generics (Generic)
import Text.PrettyPrint.HughesPJ

import Agda.Syntax.Common
import qualified Agda.Utils.Parser.MemoisedCPS as Parser

-- | Memoisation keys.

data MemoKey = NodeK      PrecedenceKey
             | PostLeftsK PrecedenceKey
             | PreRightsK PrecedenceKey
             | TopK
             | AppK
             | NonfixK
  deriving (Eq, Show, Generic)

type PrecedenceKey = Either PrecedenceLevel PrecedenceLevel

instance Hashable MemoKey

-- | The parser monad.

type Parser tok a =
#ifdef DEBUG
  Parser.ParserWithGrammar
#else
  Parser.Parser
#endif
    MemoKey tok (MaybePlaceholder tok) a

-- | Runs the parser.

parse :: forall tok a. Parser tok a -> [MaybePlaceholder tok] -> [a]
parse = Parser.parse

-- | Parses a token satisfying the given predicate. The computed value
-- is returned.

sat' :: (MaybePlaceholder tok -> Maybe a) -> Parser tok a
sat' = Parser.sat'

-- | Parses a token satisfying the given predicate.

sat :: (MaybePlaceholder tok -> Bool) ->
       Parser tok (MaybePlaceholder tok)
sat = Parser.sat

-- | Uses the given document as the printed representation of the
-- given parser. The document's precedence is taken to be 'atomP'.

doc :: Doc -> Parser tok a -> Parser tok a
doc = Parser.doc

-- | Memoises the given parser.
--
-- Every memoised parser must be annotated with a /unique/ key.
-- (Parametrised parsers must use distinct keys for distinct inputs.)

memoise :: MemoKey -> Parser tok tok -> Parser tok tok
memoise = Parser.memoise

-- | Memoises the given parser, but only if printing, not if parsing.
--
-- Every memoised parser must be annotated with a /unique/ key.
-- (Parametrised parsers must use distinct keys for distinct inputs.)

memoiseIfPrinting :: MemoKey -> Parser tok tok -> Parser tok tok
memoiseIfPrinting = Parser.memoiseIfPrinting

-- | Tries to print the parser, or returns 'empty', depending on the
-- implementation. This function might not terminate.

grammar :: Parser tok a -> Doc
grammar = Parser.grammar
