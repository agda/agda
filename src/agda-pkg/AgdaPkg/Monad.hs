{-# LANGUAGE GeneralizedNewtypeDeriving #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE StandaloneDeriving #-}
module AgdaPkg.Monad where

import Data.Set (Set)
import qualified Data.Set as Set

import Control.Monad
import Control.Monad.Trans
import Control.Monad.Reader
import Control.Monad.Catch
import Control.Monad.State

import Agda.Packaging.Base

import AgdaPkg.PackageDesc.Base

type PkgM = PkgT IO

data PackageBuildEnv
  = PackageBuildEnv
  { pbePackage :: PackageDesc
  , pbeBuildDir :: FilePath
  , pbePkgDBs :: PkgDBStack
  , pbeAgdaOpts :: [String]
  }


newtype PkgT m a = PkgT { unPkgT :: ReaderT PackageBuildEnv m a}
  deriving (MonadIO, Monad, Applicative, Functor, MonadThrow, MonadCatch, MonadMask)

runPkgT :: PackageBuildEnv -> PkgT m a -> m a
runPkgT e c = runReaderT (unPkgT c) e

class Monad m => MonadPkg m where
  getPackageBuildEnv :: m PackageBuildEnv

instance Monad m => MonadPkg (PkgT m) where
  getPackageBuildEnv = PkgT ask

instance MonadPkg m => MonadPkg (StateT s m) where
  getPackageBuildEnv = lift getPackageBuildEnv

instance MonadPkg m => MonadPkg (ReaderT e m) where
  getPackageBuildEnv = lift getPackageBuildEnv


data BuildDir = WithDir FilePath | TempDir
  deriving (Show)

data GlobalOptions
  = GlobalOptions
  { gAgdaOpts :: [String]
  , gBuildDir :: BuildDir
  , gPkgDBs :: [FilePath]
  , gPkgFilter :: PackageFilter
  }
  deriving (Show)

data PackageFilter
  = PAll
  | PNone -- filter all packages
  | PByPkgId (Set PackageId) -- only expose packages with given ids
  deriving (Show)

data Target = HtmlDoc | LaTeXDoc | CabalPackage
  deriving (Show, Eq, Ord)

type Targets = Set Target


getBuildDir :: MonadPkg m => m FilePath
getBuildDir = pbeBuildDir <$> getPackageBuildEnv

getPkgDBs :: MonadPkg m => m PkgDBStack
getPkgDBs = pbePkgDBs <$> getPackageBuildEnv

getCurrentPackage :: MonadPkg m => m PackageDesc
getCurrentPackage = pbePackage <$> getPackageBuildEnv

getAgdaOptions :: MonadPkg m => m [String]
getAgdaOptions = pbeAgdaOpts <$> getPackageBuildEnv
