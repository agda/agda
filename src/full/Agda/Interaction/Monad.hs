{-# OPTIONS_GHC -Wunused-imports #-}

{-# OPTIONS_GHC -fwarn-orphans #-}
{-# LANGUAGE CPP               #-}

module Agda.Interaction.Monad
  ( IM
  , runIM
  , readline
  ) where

import Agda.TypeChecking.Monad
  ( HasOptions
  , MonadTCEnv
  , MonadTCM
  , MonadTCState
  , ReadTCState
  , TCErr
  , TCM, TCMT(..)
  , mapTCMT
  )
import Control.Exception (throwIO)
import Control.Monad.Except (MonadError (..))
import Control.Monad.Trans (MonadIO, lift, liftIO)
import qualified System.Console.Haskeline as Haskeline

-- MonadException is replaced by MonadCatch in haskeline 0.8
#if MIN_VERSION_haskeline(0,8,0)
import qualified Control.Monad.Catch as Haskeline (catch)
#endif

-- | Interaction monad.
newtype IM a = IM {unIM :: TCMT (Haskeline.InputT IO) a}
  deriving
  ( Functor, Applicative, Monad, MonadIO
  , HasOptions, MonadTCEnv, ReadTCState, MonadTCState, MonadTCM )

runIM :: IM a -> TCM a
runIM = mapTCMT (Haskeline.runInputT Haskeline.defaultSettings) . unIM

instance MonadError TCErr IM where
  throwError                = liftIO . throwIO
  catchError (IM (TCM m)) h = IM . TCM $ \s e ->
    m s e `Haskeline.catch` \err -> unTCM (unIM (h err)) s e

-- | Line reader. The line reader history is not stored between
-- sessions.
readline :: String -> IM (Maybe String)
readline s = IM $ lift (Haskeline.getInputLine s)
