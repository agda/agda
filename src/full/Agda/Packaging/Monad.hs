{-# OPTIONS_GHC -Wunused-imports #-}

module Agda.Packaging.Monad where

{-
  ( module Control.Monad.Reader
  , AgdaPkg (..)
  , asksM ) where

-- Standard Library Imports
import Control.Applicative
import Control.Monad.Error
import Control.Monad.Reader

-- Local Imports
import Agda.Packaging.Config

--------------------------------------------------------------------------------

-- Parametric in `opt' only so that the environment can be decoupled
-- from the concrete CLI tool
newtype AgdaPkg opt a
  =     AgdaPkg
  { runAgdaPkg :: ReaderT (AgdaPkgConfig opt) IO a }
  deriving
    ( Functor
    , Monad
    , MonadError IOError
    , MonadReader (AgdaPkgConfig opt)
    , MonadIO )

instance Applicative (AgdaPkg opt) where
  pure  = return
  (<*>) = ap

asksM :: (MonadReader r m) => (r -> m a) -> m a
asksM = join . asks
-}
