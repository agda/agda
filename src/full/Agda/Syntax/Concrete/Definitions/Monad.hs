{-# OPTIONS_GHC -Wunused-imports #-}

module Agda.Syntax.Concrete.Definitions.Monad where

import Prelude hiding ( null )

import Control.Monad        ( unless )
import Control.Monad.Except ( MonadError(..), ExceptT, runExceptT )
import Control.Monad.Reader ( MonadReader, ReaderT, runReaderT )
import Control.Monad.State  ( MonadState(..), modify, State, runState )

import Data.Bifunctor (second)
import Data.Map (Map)
import qualified Data.Map as Map

import Agda.Syntax.Position
import Agda.Syntax.Common hiding (TerminationCheck())
import Agda.Syntax.Concrete.Name
import Agda.Syntax.Concrete.Definitions.Types
import Agda.Syntax.Concrete.Definitions.Errors

import Agda.Utils.CallStack ( CallStack, HasCallStack, withCallerCallStack )
import Agda.Utils.Lens
import qualified Agda.Utils.List1 as List1
import Agda.Utils.Null (Null(..))

import Agda.Utils.Impossible

-- | Nicifier monad.
--   Preserve the state when throwing an exception.

newtype Nice a = Nice { unNice :: ReaderT NiceEnv (ExceptT DeclarationException (State NiceState)) a }
  deriving ( Functor, Applicative, Monad
           , MonadReader NiceEnv, MonadState NiceState, MonadError DeclarationException
           )

-- | Run a Nicifier computation, return result and warnings
--   (in chronological order).
runNice :: NiceEnv -> Nice a -> (Either DeclarationException a, NiceWarnings)
runNice env m = second (reverse . niceWarn) $
  runExceptT (unNice m `runReaderT` env) `runState` initNiceState

instance Null a => Null (Nice a) where
  empty = pure empty
  null _ = __IMPOSSIBLE__

-- | Nicifier parameters.

data NiceEnv = NiceEnv
  { safeButNotBuiltin :: Bool
       -- ^ We are in a module declared @--safe@ which is not a builtin module.
  }

-- | Nicifier state.

data NiceState = NiceState
  { _loneSigs :: LoneSigs
    -- ^ Lone type signatures that wait for their definition.
  , _termChk  :: TerminationCheck
    -- ^ Termination checking pragma waiting for a definition.
  , _posChk   :: PositivityCheck
    -- ^ Positivity checking pragma waiting for a definition.
  , _uniChk   :: UniverseCheck
    -- ^ Universe checking pragma waiting for a data/rec signature or definition.
  , _catchall :: Catchall
    -- ^ Catchall pragma waiting for a function clause.
  , _covChk  :: CoverageCheck
    -- ^ Coverage pragma waiting for a definition.
  , niceWarn :: NiceWarnings
    -- ^ Stack of warnings. Head is last warning.
  , _nameId  :: NameId
    -- ^ We distinguish different 'NoName's (anonymous definitions) by a unique 'NameId'.
  }

data LoneSig = LoneSig
  { loneSigRange :: Range
  , loneSigName  :: Name
      -- ^ If 'isNoName', this name can have a different 'NameId'
      --   than the key of 'LoneSigs' pointing to it.
  , loneSigKind  :: DataRecOrFun
  }
  deriving Show

type LoneSigs     = Map Name LoneSig
     -- ^ We retain the 'Name' also in the codomain since
     --   'Name' as a key is up to @Eq Name@ which ignores the range.
     --   However, without range names are not unique in case the
     --   user gives a second definition of the same name.
     --   This causes then problems in 'replaceSigs' which might
     --   replace the wrong signature.
     --
     --   Another reason is that we want to distinguish different
     --   occurrences of 'NoName' in a mutual block (issue #4157).
     --   The 'NoName' in the codomain will have a unique 'NameId'.

type NiceWarnings = [DeclarationWarning]
     -- ^ Stack of warnings. Head is last warning.

-- | Initial nicifier state.

initNiceState :: NiceState
initNiceState = NiceState
  { _loneSigs = Map.empty
  , _termChk  = TerminationCheck
  , _posChk   = YesPositivityCheck
  , _uniChk   = YesUniverseCheck
  , _catchall = False
  , _covChk   = YesCoverageCheck
  , niceWarn  = []
  , _nameId   = NameId 1 noModuleNameHash
  }

lensNameId :: Lens' NiceState NameId
lensNameId f e = f (_nameId e) <&> \ i -> e { _nameId = i }

nextNameId :: Nice NameId
nextNameId = do
  i <- use lensNameId
  lensNameId %= succ
  return i

-- * Handling the lone signatures, stored to infer mutual blocks.

-- | Lens for field '_loneSigs'.

loneSigs :: Lens' NiceState LoneSigs
loneSigs f e = f (_loneSigs e) <&> \ s -> e { _loneSigs = s }

-- | Adding a lone signature to the state.
--   Return the name (which is made unique if 'isNoName').

addLoneSig :: Range -> Name -> DataRecOrFun -> Nice Name
addLoneSig r x k = do
  -- Andreas, 2020-05-19, issue #4157, make '_' unique.
  x' <- case x of
    Name{}     -> pure x
    NoName r _ -> NoName r <$> nextNameId
  loneSigs %== \ s -> do
    let (mr, s') = Map.insertLookupWithKey (\ _k new _old -> new) x (LoneSig r x' k) s
    case mr of
      Nothing -> return s'
      Just{}  -> declarationException $
        if not $ isNoName x then DuplicateDefinition x else DuplicateAnonDeclaration r
  return x'

-- | Remove a lone signature from the state.

removeLoneSig :: Name -> Nice ()
removeLoneSig x = loneSigs %= Map.delete x

-- | Search for forward type signature.

getSig :: Name -> Nice (Maybe DataRecOrFun)
getSig x = fmap loneSigKind . Map.lookup x <$> use loneSigs

-- | Check that no lone signatures are left in the state.

noLoneSigs :: Nice Bool
noLoneSigs = null <$> use loneSigs

forgetLoneSigs :: Nice ()
forgetLoneSigs = loneSigs .= Map.empty

-- | Ensure that all forward declarations have been given a definition.
checkLoneSigs :: LoneSigs -> Nice ()
checkLoneSigs xs = do
  forgetLoneSigs
  unless (Map.null xs) $ declarationWarning $ MissingDefinitions $
    map (\s -> (loneSigName s , loneSigRange s)) $ Map.elems xs

-- | Ensure that all forward declarations have been given a definition,
-- raising an error indicating *why* they would have had to have been
-- defined.
breakImplicitMutualBlock :: Range -> String -> Nice ()
breakImplicitMutualBlock r why = do
  m <- use loneSigs
  List1.unlessNull (Map.elems m) $ \ xs ->
    declarationException $ DisallowedInterleavedMutual r why $
      -- Andreas, 2023-09-07: We discard the 'loneSigRange's because the 'Name' already has a range.
      fmap loneSigName xs

-- | Get names of lone function signatures, plus their unique names.

loneFuns :: LoneSigs -> [(Name,Name)]
loneFuns = map (second loneSigName) . filter (isFunName . loneSigKind . snd) . Map.toList

-- | Create a 'LoneSigs' map from an association list.

loneSigsFromLoneNames :: [(Range, Name, DataRecOrFun)] -> LoneSigs
loneSigsFromLoneNames = Map.fromListWith __IMPOSSIBLE__ . map (\(r,x,k) -> (x, LoneSig r x k))

-- | Lens for field '_termChk'.

terminationCheckPragma :: Lens' NiceState TerminationCheck
terminationCheckPragma f e = f (_termChk e) <&> \ s -> e { _termChk = s }

withTerminationCheckPragma :: TerminationCheck -> Nice a -> Nice a
withTerminationCheckPragma tc f = do
  tc_old <- use terminationCheckPragma
  terminationCheckPragma .= tc
  result <- f
  terminationCheckPragma .= tc_old
  return result

coverageCheckPragma :: Lens' NiceState CoverageCheck
coverageCheckPragma f e = f (_covChk e) <&> \ s -> e { _covChk = s }

withCoverageCheckPragma :: CoverageCheck -> Nice a -> Nice a
withCoverageCheckPragma tc f = do
  tc_old <- use coverageCheckPragma
  coverageCheckPragma .= tc
  result <- f
  coverageCheckPragma .= tc_old
  return result

-- | Lens for field '_posChk'.

positivityCheckPragma :: Lens' NiceState PositivityCheck
positivityCheckPragma f e = f (_posChk e) <&> \ s -> e { _posChk = s }

withPositivityCheckPragma :: PositivityCheck -> Nice a -> Nice a
withPositivityCheckPragma pc f = do
  pc_old <- use positivityCheckPragma
  positivityCheckPragma .= pc
  result <- f
  positivityCheckPragma .= pc_old
  return result

-- | Lens for field '_uniChk'.

universeCheckPragma :: Lens' NiceState UniverseCheck
universeCheckPragma f e = f (_uniChk e) <&> \ s -> e { _uniChk = s }

withUniverseCheckPragma :: UniverseCheck -> Nice a -> Nice a
withUniverseCheckPragma uc f = do
  uc_old <- use universeCheckPragma
  universeCheckPragma .= uc
  result <- f
  universeCheckPragma .= uc_old
  return result

-- | Get universe check pragma from a data/rec signature.
--   Defaults to 'YesUniverseCheck'.

getUniverseCheckFromSig :: Name -> Nice UniverseCheck
getUniverseCheckFromSig x = maybe YesUniverseCheck universeCheck <$> getSig x

-- | Lens for field '_catchall'.

catchallPragma :: Lens' NiceState Catchall
catchallPragma f e = f (_catchall e) <&> \ s -> e { _catchall = s }

-- | Get current catchall pragma, and reset it for the next clause.

popCatchallPragma :: Nice Catchall
popCatchallPragma = do
  ca <- use catchallPragma
  catchallPragma .= False
  return ca

withCatchallPragma :: Catchall -> Nice a -> Nice a
withCatchallPragma ca f = do
  ca_old <- use catchallPragma
  catchallPragma .= ca
  result <- f
  catchallPragma .= ca_old
  return result

-- | Add a new warning.
niceWarning :: DeclarationWarning -> Nice ()
niceWarning w = modify $ \ st -> st { niceWarn = w : niceWarn st }

declarationException :: HasCallStack => DeclarationException' -> Nice a
declarationException e = withCallerCallStack $ throwError . flip DeclarationException e

declarationWarning' :: DeclarationWarning' -> CallStack -> Nice ()
declarationWarning' w loc = niceWarning $ DeclarationWarning loc w

declarationWarning :: HasCallStack => DeclarationWarning' -> Nice ()
declarationWarning = withCallerCallStack . declarationWarning'
