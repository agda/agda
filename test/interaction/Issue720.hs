{-# LANGUAGE RecordWildCards #-}

import Control.Monad
import System.Directory

import RunAgda
import Agda.Version

file          = "Issue720.agda"
interfaceFile = concat [ "_build/", version, "/agda/", file, "i" ]

main :: IO ()
main = runAgda ["--no-libraries"] $ \(AgdaCommands { .. }) -> do

  let load = do
        send $ command "load" file
                       (Just "Interactive Direct")
                       (Just $ show file ++ " []")
        echoUntilPrompt

  -- Discard the first prompt.
  echoUntilPrompt

  -- Load the file, and wait for Agda to complete.
  load

  -- Remove the interface file.
  exists <- doesFileExist interfaceFile
  when exists $ removeFile interfaceFile

  -- Reload.
  load

  return ()
