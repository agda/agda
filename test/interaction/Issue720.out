(agda2-status-action "")
(agda2-info-action "*Type-checking*" "" nil)
(agda2-highlight-clear)
(agda2-info-action "*Type-checking*" "Checking Issue720 (Issue720.agda)." t)
(agda2-highlight-add-annotations 'nil '(1 7 (keyword) t) '(17 22 (keyword) t))
(agda2-highlight-add-annotations 'nil '(1 7 (keyword) t) '(8 16 (module) nil nil ("Issue720.agda" . 1)) '(17 22 (keyword) t))
(agda2-highlight-add-annotations 'nil '(1 7 (keyword) t) '(8 16 (module) nil nil ("Issue720.agda" . 1)) '(17 22 (keyword) t))
(agda2-status-action "Checked")
(agda2-info-action "*All Done*" "" nil)
((last . 1) . (agda2-goals-action '()))
(agda2-status-action "")
(agda2-info-action "*Type-checking*" "" nil)
(agda2-highlight-clear)
(agda2-info-action "*Type-checking*" "Checking Issue720 (Issue720.agda)." t)
(agda2-highlight-add-annotations 'nil '(1 7 (keyword) t) '(17 22 (keyword) t))
(agda2-highlight-add-annotations 'nil '(1 7 (keyword) t) '(8 16 (module) nil nil ("Issue720.agda" . 1)) '(17 22 (keyword) t))
(agda2-highlight-add-annotations 'nil '(1 7 (keyword) t) '(8 16 (module) nil nil ("Issue720.agda" . 1)) '(17 22 (keyword) t))
(agda2-status-action "Checked")
(agda2-info-action "*All Done*" "" nil)
((last . 1) . (agda2-goals-action '()))
